pkg/pdp/service/piece_upload.go

```diff
@@ -42,175 +26,42 @@ func (p *PDPService) UploadPiece(ctx context.Context, pieceUpload types.PieceUpl
 		return types.WrapError(types.KindInternal, "failed to query for piece upload", err)
 	}
 
-	// PieceRef is a pointer, so a nil value means it's NULL in the DB.
-	if upload.PieceRef != nil {
-		return types.NewError(types.KindInvalidInput, "piece upload already uploaded")
-	}
-
-	ph := types.Piece{
-		Name: upload.CheckHashCodec,
-		Hash: hex.EncodeToString(upload.CheckHash),
-		Size: upload.CheckSize,
-	}
-	phMh, err := Multihash(ph)
-	if err != nil {
-		return err
-	}
-
-	// Limit the size of the piece data
-	maxPieceSize := upload.CheckSize
-
-	// Create a commp.Calc instance for calculating commP
-	// NB(forrest): calculation of the commp hash is the bottleneck of the upload process.
-	// throughput of the commp hash function on a modern machine is ~1.5Gbps, or less on older
-	// machines without a CPU using sha extensions.
-	// The uploader supplying the reader that is `pieceUpload.Data` may only send at the rate
-	// this host can hash the data.
-	cp := &commp.Calc{}
-	readSize := int64(0)
-
-	var vhash hash.Hash
-	if upload.CheckHashCodec != multihash.Codes[multihash.SHA2_256_TRUNC254_PADDED] {
-		hasher, err := mhreg.GetVariableHasher(multihash.Names[upload.CheckHashCodec], -1)
+	var hasher hash.Hash
+	// TODO where can I get a better value than this shitty string?
+	commpHashCodec := "fr32-sha2-256-trunc254-padded-binary-tree"
+	if upload.CheckHashCodec == commpHashCodec {
```

Comment:
`doCommp` also allows `multicodec.Sha2_256Trunc254Padded` however, it will be rejected here.

Reply:
Implemented a hash registry with a set of supported hash function as you suggested in another comment.